editor/scene_tree_dock.cpp

```diff
@@ -2038,17 +2038,155 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
 	}
 }
 
+void SceneTreeDock::_update_all_gizmos(Node *p_node) {
+	Node3DEditor::get_singleton()->update_all_gizmos(p_node);
+}
+
+void SceneTreeDock::_reparent_nodes_to_root(Node *p_root, Array p_nodes, Node *p_owner) {
+	List<Node *> nodes;
+	for (int i = 0; i < p_nodes.size(); i++) {
+		Node *node = Object::cast_to<Node>(p_nodes[i]);
+		ERR_FAIL_COND(node == nullptr);
+		nodes.push_back(node);
+	}
+
+	for (Node *node : nodes) {
+		node->set_owner(p_owner);
+		if (node != p_root && node->get_owner() == p_owner) {
+			List<Node *> owned;
+			node->get_owned_by(p_owner, &owned, true);
+			// For Node2D and Node3D types, we can preserve the global transform when move them to the root node
+			if (Object::cast_to<Node3D>(p_root)) {
+				Node3D *node_3d = Object::cast_to<Node3D>(node);
+				if (node_3d) {
+					Transform3D gt = node_3d->get_global_transform();
+					node_3d->get_parent()->remove_child(node_3d);
+					p_root->add_child(node_3d);
+					node_3d->set_global_transform(gt);
+				} else {
+					node->get_parent()->remove_child(node);
+					p_root->add_child(node);
+				}
+			} else if (Object::cast_to<Node2D>(p_root)) {
+				Node2D *node_2d = Object::cast_to<Node2D>(node);
+				if (node_2d) {
+					Transform2D gt = node_2d->get_global_transform();
+					node_2d->get_parent()->remove_child(node_2d);
+					p_root->add_child(node_2d);
+					node_2d->set_global_transform(gt);
+				} else {
+					node->get_parent()->remove_child(node);
+					p_root->add_child(node);
+				}
+			} else {
+				node->get_parent()->remove_child(node);
+				p_root->add_child(node);
+			}
```

Comment:
This can be replaced by the newly added `Node.reparent()`, which handles transforms.